File: src/sdk/go/appmesh_client_tcp.go
package appmesh

import (
	"encoding/binary"
	"errors"
	"html"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"path"
	"strings"
	"sync"

	"github.com/rs/xid"
)

const TCP_MESSAGE_HEADER_LENGTH = 4
const TCP_CHUNK_BLOCK_SIZE = 16*1024 - 256 // target to 16KB

// AppMeshClientTCP interacts with the TCP server using REST API requests via a socket.
type AppMeshClientTCP struct {
	*AppMeshClient
	TcpExecutor *ClientRequesterTcp // used for file operate functions
}

// NewTcpClient creates a new AppMeshClientTCP instance for interacting with a TCP server.
func NewTcpClient(options Option) (*AppMeshClientTCP, error) {
	tcpRequester := &ClientRequesterTcp{
		BaseURL: func() string {
			if options.AppMeshUri != "" {
				return options.AppMeshUri
			}
			return DEFAULT_TCP_URI
		}(),
	}
	options.tcpOnly = new(bool)

	client := &AppMeshClientTCP{AppMeshClient: NewHttpClient(options), TcpExecutor: tcpRequester}

	client.AppMeshClient.Proxy = tcpRequester

	return client, tcpRequester.connectSocket(
		client.sslClientCert,
		client.sslClientCertKey,
		client.sslCAFile,
	)
}

// CloseConnection closes the TCP connection
func (client *AppMeshClientTCP) CloseConnection() {
	if client.TcpExecutor != nil && client.TcpExecutor.SocketConn != nil {
		client.TcpExecutor.SocketConn.Close()
		client.TcpExecutor.SocketConn = nil
	}
}

// FileDownload downloads a file from the server to the local file system
func (r *AppMeshClientTCP) FileDownload(remoteFile, localFile string, applyFileAttributes bool) error {
	headers := map[string]string{
		"File-Path":                        remoteFile,
		HTTP_HEADER_KEY_X_RECV_FILE_SOCKET: "true",
	}

	status, msg, responseHeaders, err := r.get("/appmesh/file/download", nil, headers)
	if err != nil || status != http.StatusOK {
		return errors.New("failed to download file: " + string(msg))
	}

	if _, ok := responseHeaders[HTTP_HEADER_KEY_X_RECV_FILE_SOCKET]; !ok {
		return errors.New("HTTP_HEADER_KEY_X_RECV_FILE_SOCKET header missing")
	}

	return r.receiveFile(localFile, responseHeaders, applyFileAttributes)
}

// Helper function to receive file data
func (r *AppMeshClientTCP) receiveFile(localFile string, headers http.Header, applyFileAttributes bool) error {
	file, err := os.Create(localFile)
	if err != nil {
		return err
	}
	defer file.Close()

	for {
		chunkSizeBuf, err := r.TcpExecutor.readTcpData(TCP_MESSAGE_HEADER_LENGTH)
		if err != nil {
			return err
		}

		chunkSize := binary.BigEndian.Uint32(chunkSizeBuf)
		if chunkSize == 0 {
			break
		}

		chunkData, err := r.TcpExecutor.readTcpData((chunkSize))
		if err != nil {
			return err
		}

		if _, err := file.Write(chunkData); err != nil {
			return err
		}
	}

	if applyFileAttributes {
		SetFileAttributes(localFile, headers)
	}
	return nil
}

// FileUpload uploads a local file to the server
func (client *AppMeshClientTCP) FileUpload(localFile, remoteFile string, applyFileAttributes bool) error {
	file, err := os.Open(localFile)
	if err != nil {
		return err
	}
	defer file.Close()

	headers := map[string]string{
		"File-Path":                        remoteFile,
		"Content-Type":                     "text/plain",
		HTTP_HEADER_KEY_X_SEND_FILE_SOCKET: "true",
	}

	// Get the file attributes
	if applyFileAttributes {
		attrs, err := GetFileAttributes(localFile)
		MergeStringMaps(headers, attrs)
		if err != nil {
			return err
		}
	}

	status, _, responseHeaders, err := client.post("/appmesh/file/upload", nil, headers, nil)
	if err != nil || status != http.StatusOK {
		return errors.New("failed to upload file")
	}

	if responseHeaders.Get(HTTP_HEADER_KEY_X_SEND_FILE_SOCKET) == "" {
		return errors.New("HTTP_HEADER_KEY_X_SEND_FILE_SOCKET header missing")
	}
	return client.uploadFileChunks(file)
}

// Helper function to upload file in chunks
func (client *AppMeshClientTCP) uploadFileChunks(file *os.File) error {
	chunkSize := TCP_CHUNK_BLOCK_SIZE
	buffer := make([]byte, chunkSize)

	for {
		n, err := file.Read(buffer)
		if err != nil && err != io.EOF {
			return err
		}
		if n == 0 {
			break
		}

		if err := client.TcpExecutor.SendData(buffer[:n]); err != nil {
			return err
		}
	}

	// Send end-of-file marker
	return client.TcpExecutor.sendEOFMarker()
}

// TCP Request executor
type ClientRequesterTcp struct {
	BaseURL    string
	SocketConn net.Conn
	mutex      sync.Mutex
}

// doRequest performs a REST-like request over TCP
func (r *ClientRequesterTcp) doRequest(method, apiPath string, queries url.Values, headers map[string]string, body io.Reader, token string, forwardingHost string) (int, []byte, http.Header, error) {
	u, _ := ParseURL(r.BaseURL)
	u.Path = path.Join(u.Path, apiPath)
	if queries != nil {
		u.RawQuery = queries.Encode()
	}
	req, err := http.NewRequest(method, u.String(), body)
	if err != nil {
		return 0, nil, nil, err
	}

	// Set headers
	if token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	if forwardingHost != "" {
		if strings.Contains(forwardingHost, ":") {
			req.Header.Set("X-Target-Host", forwardingHost)
		} else {
			req.Header.Set("X-Target-Host", forwardingHost+":"+u.Port())
		}
	}
	req.Header.Set(HTTP_USER_AGENT_HEADER_NAME, HTTP_USER_AGENT_TCP)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")
	for k, v := range headers {
		req.Header.Add(k, v)
	}

	// Execute the TCP request
	resp, err := r.request(req)

	// Format the response headers
	respHeaders := make(http.Header)
	for key, value := range resp.Headers {
		respHeaders.Add(key, value)
	}

	return resp.HttpStatus, []byte(resp.Body), respHeaders, err
}

// connectSocket establishes a secure TCP connection
func (r *ClientRequesterTcp) connectSocket(sslClientCert string, sslClientCertKey string, sslCAFile string) error {
	u, err := ParseURL(r.BaseURL)
	if err != nil {
		return err
	}
	address := net.JoinHostPort(u.Hostname(), u.Port())

	ssl := &SSLConfig{
		VerifyClient:                sslClientCert != "" && sslClientCertKey != "",
		SSLClientCertificateFile:    sslClientCert,
		SSLClientCertificateKeyFile: sslClientCertKey,
		VerifyServer:                sslCAFile != "",
		SSLCaPath:                   sslCAFile,
	}

	r.SocketConn, err = ConnectAppMeshServer(address, ssl.VerifyServer, ssl)
	return err
}

// request sends a request over TCP
func (r *ClientRequesterTcp) request(req *http.Request) (*Response, error) {

	data := new(Request)
	data.Uuid = xid.New().String()
	data.HttpMethod = req.Method
	data.RequestUri = req.URL.Path
	data.ClientAddress = r.SocketConn.LocalAddr().String()
	data.Headers = make(map[string]string)
	for key, values := range req.Header {
		for _, value := range values {
			data.Headers[string(key)] = string(value)
		}
	}
	data.Queries = make(map[string]string)
	queryParams := req.URL.Query()
	for key, values := range queryParams {
		for _, value := range values {
			data.Queries[string(key)] = string(value)
		}
	}

	// do not read body for file upload
	if !(data.HttpMethod != "POST" && data.RequestUri != "/appmesh/file/upload") {
		if bodyBytes, err := io.ReadAll(req.Body); err == nil && len(bodyBytes) > 0 {
			data.Body = html.UnescapeString(string(bodyBytes))
		}
	}

	// Set user agent
	data.Headers[HTTP_USER_AGENT_HEADER_NAME] = HTTP_USER_AGENT_TCP

	// Serialize the request
	buf, err := data.Serialize()
	if err != nil {
		return nil, err
	}

	// Send the data over TCP
	if err := r.SendData(buf); err != nil {
		return nil, err
	}

	// Receive the response
	respData, err := r.RecvData()
	if err != nil {
		return nil, err
	}

	// Deserialize the response
	respMsg := &Response{}
	if err := respMsg.Deserialize(respData); err != nil {
		return nil, err
	}
	return respMsg, nil
}

// SendData sends the message length and the actual message data over TCP
func (r *ClientRequesterTcp) SendData(data []byte) error {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	// Send length header
	if err := r.sendLength(len(data)); err != nil {
		return err
	}

	// Send message data
	return r.sendTcpData(data)
}

// sendLength sends the length of the data as a 4-byte header
func (r *ClientRequesterTcp) sendLength(length int) error {
	lenBuf := make([]byte, TCP_MESSAGE_HEADER_LENGTH)
	binary.BigEndian.PutUint32(lenBuf, uint32(length))
	return r.sendTcpData(lenBuf)
}

func (r *ClientRequesterTcp) sendTcpData(buf []byte) error {
	var totalSentSize int = 0
	var err error = nil
	for totalSentSize < len(buf) {
		byteSent := 0

		if byteSent, err = r.SocketConn.Write(buf[totalSentSize:]); err != nil {
			return err
		}
		totalSentSize += byteSent
		//log.Printf("total send size %d bytes", totalSentSize)
	}
	return err
}

// sendEOFMarker sends an end-of-file marker
func (r *ClientRequesterTcp) sendEOFMarker() error {
	return r.sendLength(0)
}

// RecvData receives a length-prefixed message from the TCP connection
func (r *ClientRequesterTcp) RecvData() ([]byte, error) {
	lenBuf, err := r.readTcpData(TCP_MESSAGE_HEADER_LENGTH)
	if err != nil {
		return nil, err
	}

	msgLen := binary.BigEndian.Uint32(lenBuf)
	return r.readTcpData((msgLen))
}

// readTcpData reads a specific number of bytes from the TCP connection
func (r *ClientRequesterTcp) readTcpData(msgLength uint32) ([]byte, error) {
	// read body buffer
	var chunkSize uint32 = TCP_CHUNK_BLOCK_SIZE
	if msgLength < chunkSize {
		chunkSize = msgLength
	}
	// make 0 length data bytes (since we'll be appending)
	bodyBuf := make([]byte, 0)
	var totalReadSize uint32 = 0
	for totalReadSize < msgLength {
		// https://stackoverflow.com/questions/24339660/read-whole-data-with-golang-net-conn-read
		oneTimeRead := msgLength - totalReadSize
		if oneTimeRead > chunkSize {
			oneTimeRead = chunkSize
		}
		data := make([]byte, oneTimeRead) //TODO: use global buffer avoid garbage
		n, err := r.SocketConn.Read(data)
		if n > 0 {
			bodyBuf = append(bodyBuf, data[:n]...)
			totalReadSize += uint32(n)
			//log.Printf("expect: %d, read: %d, left: %d", oneTimeRead, n, msgLength-totalReadSize)
			continue
		} else if err != nil {
			return nil, err
		}
	}

	return bodyBuf, nil
}

File: src/sdk/java/src/main/java/AppMeshClientTCP.java
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.json.JSONObject;

/**
 * Simple TCP (TLS) client wrapper that reuses `AppMeshClient`.
 *
 * It constructs an HTTPS base URL using the provided TCP address and
 * reuses `AppMeshClient` for underlying HTTP requests. This mirrors the
 * Python approach of injecting a TCP client while reusing the same public API.
 */
public class AppMeshClientTCP extends AppMeshClient {
    private static final Logger LOGGER = Logger.getLogger(AppMeshClientTCP.class.getName());
    private static final int TCP_BLOCK_SIZE = 16 * 1024 - 128; // TLS-optimized chunk size
    private final TCPTransport tcpTransport;
    private volatile MessageDemuxer demuxer;

    /** Create a TCP transport client with default TLS verification behavior. */
    public AppMeshClientTCP(String host, int port) {
        this(host, port, false, null, null, null);
    }

    /**
     * Create a TCP transport client that reuses the standard App Mesh client API.
     *
     * <p>Control requests use the normal SDK surface, while file transfers switch to the TCP
     * file-socket side channel for larger payloads.
     */
    public AppMeshClientTCP(String host, int port, boolean disableSSLVerification, String caCertPath,
            String clientCertPath, String clientKeyPath) {
        super(new AppMeshClient.Builder().baseURL("https://" + host + ":" + port));
        Objects.requireNonNull(host, "host");
        try {
            javax.net.ssl.SSLContext sc = Utils.createSSLContext(caCertPath, clientCertPath, clientKeyPath,
                    disableSSLVerification);
            this.tcpTransport = new TCPTransport(host, port, sc.getSocketFactory());
        } catch (Exception e) {
            throw new RuntimeException("Failed to initialize TCP transport SSL context", e);
        }
    }

    /** Builder for {@link AppMeshClientTCP} to allow fluent configuration. */
    public static class Builder {
        private String host = "127.0.0.1";
        private int port = 6059;
        private boolean disableSSLVerification = false;
        private String caCertPath;
        private String clientCertPath;
        private String clientKeyPath;

        public Builder() {
        }

        public Builder host(String host) {
            this.host = host;
            return this;
        }

        public Builder port(int port) {
            this.port = port;
            return this;
        }

        public Builder disableSSLVerify() {
            this.disableSSLVerification = true;
            return this;
        }

        public Builder caCert(String caCertPath) {
            this.caCertPath = caCertPath;
            return this;
        }

        public Builder clientCert(String clientCertPath, String clientKeyPath) {
            this.clientCertPath = clientCertPath;
            this.clientKeyPath = clientKeyPath;
            return this;
        }

        public AppMeshClientTCP build() {
            Objects.requireNonNull(host, "host");
            return new AppMeshClientTCP(host, port, disableSSLVerification, caCertPath, clientCertPath,
                    clientKeyPath);
        }
    }

    /** Send a request over TCP transport and expose it as an {@link HttpURLConnection}-like object. */
    @Override
    public HttpURLConnection request(String method, String path, Object body, java.util.Map<String, String> headers,
            java.util.Map<String, String> params) throws IOException {
        try {
            if (!tcpTransport.connected()) {
                tcpTransport.connect();
            }

            RequestMessage req = new RequestMessage();
            req.uuid = java.util.UUID.randomUUID().toString();
            req.http_method = method;
            req.request_uri = path;
            req.client_addr = java.net.InetAddress.getLocalHost().getHostName();
            // Merge common headers from parent
            req.headers = new java.util.HashMap<>();
            req.headers.putAll(super.getCommonHeaders());
            if (headers != null)
                req.headers.putAll(headers);
            if (params != null)
                req.query = params;
            if (body != null) {
                if (body instanceof byte[])
                    req.body = (byte[]) body;
                else
                    req.body = body.toString().getBytes(StandardCharsets.UTF_8);
            }

            byte[] data = req.serialize();

            ResponseMessage resp;
            if (demuxer != null && demuxer.isRunning()) {
                // Demuxer is active: register before send to avoid race
                demuxer.registerPending(req.uuid);
                try {
                    tcpTransport.sendMessage(data);
                    resp = demuxer.waitForResponse(req.uuid, 60);
                    if (resp == null) {
                        throw new IOException("Demuxer response timeout");
                    }
                } finally {
                    demuxer.unregisterPending(req.uuid);
                }
            } else {
                // Legacy synchronous mode: send then read directly
                tcpTransport.sendMessage(data);
                byte[] respBuf = tcpTransport.receiveMessage();
                if (respBuf == null) {
                    throw new IOException("EOF from transport");
                }
                resp = ResponseMessage.deserialize(respBuf);
            }

            // Build a lightweight HttpURLConnection wrapper
            URL fake = new java.net.URI("http", null, "appmesh.local", -1, path, null, null).toURL();
            return new TcpHttpURLConnection(fake, resp);

        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException("TCP transport error", e);
        }
    }

    /**
     * Download a file through the TCP file-socket side channel.
     *
     * <p>When {@code applyFileAttributes} is true, returned POSIX metadata is applied locally on a
     * best-effort basis.
     */
    @Override
    public boolean downloadFile(String filePath, String localFile, boolean applyFileAttributes) throws IOException {
        Map<String, String> headers = new HashMap<>();
        headers.put("X-File-Path", filePath);
        headers.put("X-Recv-File-Socket", "true");

        // Initiate the request
        HttpURLConnection conn = request("GET", "/appmesh/file/download", null, headers, null);
        if (conn.getHeaderField("X-Recv-File-Socket") == null) {
            throw new IOException("Server did not respond with socket transfer option");
        }

        // Check for success status before proceeding
        if (conn.getResponseCode() != HttpURLConnection.HTTP_OK) {
            throw new IOException("Download failed: " + conn.getResponseCode());
        }

        // In TCP mode, subsequent messages on the socket are the file chunks
        try (OutputStream outputStream = new FileOutputStream(localFile)) {
            while (true) {
                byte[] chunk = tcpTransport.receiveMessage();
                if (chunk == null || chunk.length == 0) {
                    break;
                }
                outputStream.write(chunk);
            }
        } catch (Exception e) {
            throw new IOException("Failed to receive file chunks over TCP", e);
        }

        if (applyFileAttributes) {
            Utils.applyFileAttributes(localFile, conn);
        }
        return true;
    }

    /**
     * Upload a file through the TCP file-socket side channel.
     *
     * <p>When {@code preservePermissions} is true, local file metadata is sent so the server can
     * recreate permissions and ownership when supported.
     */
    @Override
    public boolean uploadFile(Object localFile, String remoteFile, boolean preservePermissions) throws IOException {
        File file;
        if (localFile instanceof String) {
            file = new File((String) localFile);
        } else if (localFile instanceof File) {
            file = (File) localFile;
        } else {
            throw new IllegalArgumentException("localFile must be a String path or a File object");
        }

        Path localPath = file.toPath();
        if (!Files.exists(localPath)) {
            throw new IOException("Local file not found: " + file.getAbsolutePath());
        }

        Map<String, String> headers = new HashMap<>();
        headers.put("X-File-Path", remoteFile);
        headers.put("Content-Type", "application/octet-stream");
        headers.put("X-Send-File-Socket", "true");

        if (preservePermissions) {
            headers.putAll(Utils.getFileAttributes(file));
        }

        // Initiate upload request
        HttpURLConnection conn = request("POST", "/appmesh/file/upload", null, headers, null);

        if (conn.getResponseCode() != HttpURLConnection.HTTP_OK) {
            throw new IOException("Upload init failed: " + conn.getResponseCode());
        }
        if (conn.getHeaderField("X-Send-File-Socket") == null) {
            throw new IOException("Server did not respond with socket transfer option");
        }

        // Send file chunks
        try (FileInputStream inputStream = new FileInputStream(file)) {
            byte[] buffer = new byte[TCP_BLOCK_SIZE];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                // If we read less than the buffer, copy it to a correctly sized array
                byte[] chunk = (bytesRead == buffer.length) ? buffer : Arrays.copyOf(buffer, bytesRead);
                tcpTransport.sendMessage(chunk);
            }
            // Send EOF signal (empty byte array)
            tcpTransport.sendMessage(null);
        } catch (Exception e) {
            throw new IOException("Failed to send file chunks over TCP", e);
        }

        return true;
    }

    /**
     * Enable the message demuxer for concurrent request-response and event routing.
     * Creates and starts the demuxer if not already running.
     */
    public synchronized void enableDemuxer() {
        if (demuxer != null && demuxer.isRunning()) {
            return;
        }
        demuxer = new MessageDemuxer(() -> tcpTransport.receiveMessage());
        demuxer.start();
    }

    /**
     * Return the message demuxer instance, or null if not enabled.
     */
    public MessageDemuxer getDemuxer() {
        return demuxer;
    }

    /**
     * Subscribe-based override for TCP transport.
     *
     * <p>Instead of polling, subscribes to STDOUT/EXIT/REMOVED events, backfills
     * output emitted before subscribe took effect, deduplicates by byte offset,
     * and waits for the process to finish or the connection to drop.
     *
     * <p>Sentinel exit codes: {@code null} = caller-side timeout,
     * {@code -1} = REMOVED before EXIT, {@code -2} = demuxer disconnected.
     */
    @Override
    public Integer waitForAsyncRun(AppRun run, boolean printStdout, int timeoutSeconds) throws Exception {
        if (run == null) {
            return null;
        }

        // EXIT_CODE_NONE signals "not yet set"; real exit codes are >= 0,
        // sentinels are -1 (REMOVED) and -2 (disconnected).
        final int EXIT_CODE_NONE = Integer.MIN_VALUE;
        final AtomicInteger exitCode = new AtomicInteger(EXIT_CODE_NONE);
        final AtomicLong deliveredUntil = new AtomicLong(0);
        final CountDownLatch done = new CountDownLatch(1);
        final Object deliverLock = new Object();

        // Event callback: routes STDOUT / EXIT / REMOVED / __disconnected__
        MessageDemuxer.EventCallback callback = (event) -> {
            switch (event.eventType) {
                case "STDOUT":
                    if (event.data != null) {
                        long pos = event.data.optLong("position", 0);
                        String output = event.data.optString("output", "");
                        deliverOutput(output, pos, deliveredUntil, deliverLock, printStdout);
                    }
                    break;
                case "EXIT":
                    int code = (event.data != null) ? event.data.optInt("exit_code", -1) : -1;
                    exitCode.compareAndSet(EXIT_CODE_NONE, code);
                    done.countDown();
                    break;
                case "REMOVED":
                    exitCode.compareAndSet(EXIT_CODE_NONE, -1);
                    done.countDown();
                    break;
                case MessageDemuxer.EVENT_TYPE_DISCONNECTED:
                    exitCode.compareAndSet(EXIT_CODE_NONE, -2);
                    done.countDown();
                    break;
                default:
                    break;
            }
        };

        JSONObject sub = this.subscribe(run.getAppName(),
                new String[] { "STDOUT", "EXIT", "REMOVED" }, callback);
        String subscriptionId = sub.optString("subscription_id", "");

        try {
            // Backfill bytes emitted before subscribe took effect
            try {
                AppOutput backfill = this.getAppOutput(run.getAppName(), 0, 0, 0,
                        run.getProcUid(), 0);
                if (backfill.httpBody != null && !backfill.httpBody.isEmpty()) {
                    deliverOutput(backfill.httpBody, 0, deliveredUntil, deliverLock, printStdout);
                }
                if (backfill.exitCode != null) {
                    exitCode.compareAndSet(EXIT_CODE_NONE, backfill.exitCode);
                    done.countDown();
                }
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "Backfill failed for " + run.getAppName(), e);
            }

            // Wait for done signal
            if (timeoutSeconds > 0) {
                done.await(timeoutSeconds, TimeUnit.SECONDS);
            } else {
                done.await();
            }
        } finally {
            // Unsubscribe
            try {
                if (!subscriptionId.isEmpty()) {
                    this.unsubscribe(subscriptionId);
                }
            } catch (Exception ignored) {
            }
            // Best-effort delete on a real exit (>= 0).
            // Sentinels (-1 REMOVED, -2 disconnected) mean the app is already gone.
            int finalCode = exitCode.get();
            if (finalCode != EXIT_CODE_NONE && finalCode >= 0) {
                try {
                    this.deleteApp(run.getAppName());
                } catch (Exception ignored) {
                }
            }
        }

        int result = exitCode.get();
        return (result == EXIT_CODE_NONE) ? null : result;
    }

    /**
     * Deliver stdout output with deduplication by byte offset.
     */
    private static void deliverOutput(String chunk, long pos, AtomicLong deliveredUntil,
            Object lock, boolean printStdout) {
        if (chunk == null || chunk.isEmpty()) {
            return;
        }
        byte[] chunkBytes = chunk.getBytes(StandardCharsets.UTF_8);
        synchronized (lock) {
            long current = deliveredUntil.get();
            long end = pos + chunkBytes.length;
            if (end <= current) {
                return; // already delivered
            }
            String toPrint;
            if (pos < current) {
                // Partial overlap: trim the already-delivered prefix
                int skip = (int) (current - pos);
                toPrint = new String(chunkBytes, skip, chunkBytes.length - skip, StandardCharsets.UTF_8);
            } else {
                toPrint = chunk;
            }
            deliveredUntil.set(end);
            if (printStdout && !toPrint.isEmpty()) {
                System.out.print(toPrint);
                System.out.flush();
            }
        }
    }

    @Override
    public void close() {
        if (demuxer != null) {
            demuxer.stop();
            demuxer = null;
        }
        try {
            tcpTransport.close();
        } catch (Exception ignored) {
        }
        super.close();
    }

    // Minimal HttpURLConnection wrapper around ResponseMessage
    private static class TcpHttpURLConnection extends HttpURLConnection {
        private final ResponseMessage resp;
        // Use case-insensitive map for HTTP header lookups (RFC 7230)
        private final java.util.TreeMap<String, String> headerMap =
                new java.util.TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        private final byte[] content;

        protected TcpHttpURLConnection(URL u, ResponseMessage resp) {
            super(u);
            this.resp = resp;
            if (resp.headers != null)
                headerMap.putAll(resp.headers);
            this.content = resp.body != null ? resp.body : new byte[0];
            this.connected = true;
        }

        @Override
        public void disconnect() {
            this.connected = false;
        }

        @Override
        public boolean usingProxy() {
            return false;
        }

        @Override
        public void connect() throws IOException {
            this.connected = true;
        }

        @Override
        public int getResponseCode() throws IOException {
            return resp.http_status;
        }

        @Override
        public InputStream getInputStream() throws IOException {
            if (getResponseCode() >= 400)
                throw new IOException("HTTP error " + getResponseCode());
            return new ByteArrayInputStream(content);
        }

        @Override
        public InputStream getErrorStream() {
            try {
                if (getResponseCode() >= 400)
                    return new ByteArrayInputStream(content);
            } catch (IOException e) {
                return new ByteArrayInputStream(content);
            }
            return null;
        }

        @Override
        public String getHeaderField(String name) {
            return headerMap.get(name);
        }
    }
}
